nx/lib/nx/lin_alg/svd.ex

```diff
@@ -24,22 +24,59 @@ defmodule Nx.LinAlg.SVD do
   defn svd(input_tensor, opts \\ []) do
     validate_opts(opts)
 
+    {revectorize, target_shape, u_shape, s_shape, vt_shape} = calculate_shapes(input_tensor)
+
+    tensor =
+      if revectorize do
+        Nx.revectorize(input_tensor, [vector: :auto], target_shape: target_shape)
+      else
+        input_tensor
+      end
+
+    {u, s, vt} =
+      if Nx.all(tensor == 0) do
+        svd_all_zeros(tensor, opts)
+      else
+        svd_non_zero(tensor, opts)
+      end
+
+    # we can force [] as the vectorized axes because we are guaranteed that the input is devectorized
     result =
-      if Nx.all(input_tensor == 0) do
-        svd_all_zeros(input_tensor, opts)
+      if revectorize do
+        {
+          Nx.revectorize(u, [], target_shape: u_shape),
+          Nx.revectorize(s, [], target_shape: s_shape),
+          Nx.revectorize(vt, [], target_shape: vt_shape)
+        }
       else
-        svd_non_zero(input_tensor, opts)
+        {u, s, vt}
       end
 
     custom_grad(result, [input_tensor], fn g ->
-      grad(result, input_tensor, g)
+      svd_grad(result, input_tensor, g)
     end)
   end
 
   deftransformp validate_opts(opts \\ []) do
     opts[:max_iter] || raise ArgumentError, "missing option :max_iter"
   end
 
+  deftransformp calculate_shapes(t) do
+    shape = Nx.shape(t)
+    rank = tuple_size(shape)
+    m = elem(shape, rank - 2)
+    n = elem(shape, rank - 1)
+
+    collapsed_axes = shape |> Tuple.delete_at(rank - 2) |> Tuple.delete_at(rank - 2)
+
+    u_shape = collapsed_axes |> Tuple.append(m) |> Tuple.append(:auto)
+    s_shape = Tuple.append(collapsed_axes, :auto)
+    vt_shape = Tuple.append(s_shape, n)
+
+    revectorize = collapsed_axes != {}
```

Comment:
@josevalim the PR is now mostly done! The only thing for which I couldn't find a root cause is that if we remove this, there are some cases in which we get "unexpected vectorized_axes" raises in the evaluator. Could you take a look at this?